autogpt_platform/backend/backend/data/graph.py

```diff
@@ -543,20 +548,40 @@ async def get_graph(
     where_clause: AgentGraphWhereInput = {
         "id": graph_id,
     }
+
     if version is not None:
         where_clause["version"] = version
-    elif not template:
+    else:
         where_clause["isActive"] = True
 
-    # TODO: Fix hack workaround to get adding store agents to work
-    if user_id is not None and not template:
-        where_clause["userId"] = user_id
-
     graph = await AgentGraph.prisma().find_first(
         where=where_clause,
         include=AGENT_GRAPH_INCLUDE,
         order={"version": "desc"},
     )
+
+    if not graph:
+        return None
+
+    if graph.userId == user_id:
+        return GraphModel.from_db(graph, for_export) if graph else None
+
+    # If the graph is not owned by the user, we need to check if it's a store listing.
+    if not version:
+        version = graph.version
+
+    store_listing_where: StoreListingWhereInput = {
+        "agentId": graph_id,
+        "agentVersion": version,
+    }
+
+    store_listing = await StoreListing.prisma().find_first(where=store_listing_where)
+
+    # If it does not belong to the user nor is not a store listing, return None
+    if not store_listing:
+        return None
+
+    # If it is a store listing, return the graph model
```

Comment:
```suggestion
    # The Graph has to be owned by the user or a store listing.
    if graph.userId != user_id and not (await StoreListing.prisma().find_first(where={
        "agentId": graph_id,
        "agentVersion": version or graph.version,
    })):
        return None
```
I think a concise form is easier to understand.

Reply:
Btw don't we put model queries for StoreListing in store/db.py? can we put this function there?

Reply:
Also ["agentId", "agentVersion"] has to be indexed.

Reply:
Also does that mean the query has to filter the `isApproved` & `isDeleted` too?
I think the query needs to be placed on store/db.py with these filters ?

Reply:
I've changed it the following, keeping the not null check in there:

```python
    # The Graph has to be owned by the user or a store listing.
    if graph and graph.userId != user_id and not (await StoreListing.prisma().find_first(where={
        "agentId": graph_id,
        "agentVersion": version or graph.version,
    })):
        return None
```


> Or even further, why do we need ID for StoreListing, StoreListingVersion ?

I see your point on the `StoreListing` as we want it to be unique by agent we could just use the `agentId` as the PK

We don't force agents-version to be unique in the StoreListingVersion  table. So the creator can "modify" the store listing, by resubmitting the same agent-version with new store listing data i.e. new description etc.

I did not include a filter on `isApproved` or `isDeleted` as it didn't seem an important check at this stage, I don't want to accidentally prevent people from accessing agents that they already have been running.

easy to add in but not sure what the desired behavior is here. 
What behavior do you think we want?

Reply:
@Swiftyos if you don't add mark this as a PK you definitely need to add an index.

> We don't force agents-version to be unique in the StoreListingVersion table. So the creator can "modify" the store listing, by resubmitting the same agent-version with new store listing data i.e. new description etc.

[agentId, agentVersion] StoreListing id and [agentId, agentVersion, version] StoreListingVersion.

While talking about index:

```
model StoreListingVersion {
  @@unique([agentId, agentVersion])
  @@index([agentId, agentVersion, isApproved])
}
```
* Index is not used if unique already present
* As you said the listing version should not be unique, so the unique should be [agentId, agentVersion, version] ? This can be problematic.
* Also if we did above, might as well remove the id and make it a PK.


Same with Store agent:
```
view StoreAgent {
  listing_id            String   @id
  storeListingVersionId String
  updated_at            DateTime

  slug        String
  agent_name  String
  agent_video String?
  agent_image String[]

  featured         Boolean  @default(false)
  creator_username String
  creator_avatar   String
  sub_heading      String
  description      String
  categories       String[]
  runs             Int
  rating           Float
  versions         String[]

  @@unique([creator_username, slug])
  @@index([creator_username])
  @@index([featured])
  @@index([categories])
  @@index([storeListingVersionId])
}
```
* creator_username index is not needed since we already have unique.
* index on only 1 boolean (featured) is not a good idea, this will not be used most of the time.
* Also index on categories which is a list, I can't imagine how it will be used.
* storeListingVersionId, shouldn't this What the query is look like? I think what you want is [`listing_id`, `storeListingVersionId`] ?